### PR TITLE
Try to use schema-profunctor lib.

### PR DESCRIPTION
**Pull request by fisx**

### CODE REVIEW

File: libs/wire-api/src/Wire/API/User/Client.hs [hunk 1]
```diff
@@ -97,6 +97,92 @@ import Wire.API.Arbitrary (Arbitrary (arbitrary), GenericUniform (..), generateE
...
+    ClientSupportsLegalholdImplicitConsent
+  deriving stock (Eq, Ord, Bounded, Enum, Show, Generic)
+  deriving (Arbitrary) via (GenericUniform SupportedClientFeature)
+  deriving (ToJSON, FromJSON, Swagger.ToSchema) via S.Schema SupportedClientFeature
+
+instance S.ToSchema SupportedClientFeature where
+  schema =
+    S.named "SupportedClientFeature" $
+      S.tag _ClientSupportsLegalholdImplicitConsent S.schema
+
```
Comment (pcapriotti) [inline libs/wire-api/src/Wire/API/User/Client.hs:148]: You should use `enum` here. Something like:
```haskell
S.enum @Text "SupportedClientFeature" $
  S.element "legalhold-implicit-consent" ClientSupportsLegalholdImplicitConsent
```

File: libs/wire-api/src/Wire/API/User/Client.hs [hunk 2]
```diff
@@ -97,6 +97,92 @@ import Wire.API.Arbitrary (Arbitrary (arbitrary), GenericUniform (..), generateE
...
+  Doc.description "Hints provided by the client for the backend so it can behavior in a backwards-compatible way."
+  Doc.property "feature_list" (Doc.array typeSupportedClientFeature) $ do
+    Doc.description "Array containing all supported features."
+
+instance ToJSON SupportedClientFeatureList where
+  toJSON (SupportedClientFeatureList l) = object ["feature_list" .= l]
+
+instance FromJSON SupportedClientFeatureList where
+  parseJSON = withObject "SupportedClientFeatureList" $ \obj -> SupportedClientFeatureList <$> obj .: "feature_list"
+
```
Comment (pcapriotti) [inline libs/wire-api/src/Wire/API/User/Client.hs:185]: These can be derived automatically if you write an `S.ToSchema` instance. Something like the following should work (haven't tried compiling it though):
```haskell
S.object "SupportedClientFeatureList" $
  fromSupportedClientFeatureList S..= S.field "feature_list" S.schema
```

**Review by pcapriotti — CHANGES_REQUESTED** [review]
I have left some comments.

If it's not too much of a pain, could you also make the `Schema` import unqualified, and qualify `Aeson`? This is the way I've dealt with imports in other places.

**Comment by fisx** [comment]
thanks, this is helpful!  I tried again, and got stuck a little further down, but i think i'm almost there.  :)

as for the unqualified import of `Data.Schema`, I would like to leave that for a dedicated refactoring PR that transforms the entire module, otherwise it'll get too noisy.

NB: we've tried to establish a policy in the backend chapter that forbids unqualified, implicit imports.  with limited success, admittedly, but there appeared to be a weak consensus when we debated it.

File: libs/wire-api/src/Wire/API/User/Client.hs [hunk 3]
```diff
@@ -97,6 +100,93 @@ import Wire.API.Arbitrary (Arbitrary (arbitrary), GenericUniform (..), generateE
...
+data SupportedClientFeatureList = SupportedClientFeatureList {fromSupportedClientFeatureList :: Set SupportedClientFeature}
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving (Arbitrary) via (GenericUniform SupportedClientFeatureList)
+  deriving (ToJSON, FromJSON, Swagger.ToSchema) via (Schema.Schema SupportedClientFeatureList)
+
+instance Schema.ToSchema SupportedClientFeatureList where
+  schema =
+    Schema.object "SupportedClientFeatureList" $
+      SupportedClientFeatureList
+        <$> (fromSupportedClientFeatureList Schema..= Schema.field "feature_list" Schema.schema)
```
Comment (pcapriotti) [inline libs/wire-api/src/Wire/API/User/Client.hs:179]: The brackets here should not be necessary.

**Review by pcapriotti — APPROVED** [review]
Looks good! Agreed about the unqualified import, that can be done separately.

You can also attach descriptions using `objectWithDocModifier "name" (description ?~ "desc")` instead of `object "name"`.

Minor nitpick follows.

**Comment by fisx** [comment]
looks good, but doesn't compile yet!  :)  can you guess the type error from looking at the change?

**Comment by fisx** [comment]
spoiler:

```
/home/mf/src/wire/wire-server/libs/wire-api/src/Wire/API/User/Client.hs:182:12: error:
    • Couldn't match type ‘[SupportedClientFeature]’
                     with ‘Set SupportedClientFeature’
```

**Comment by pcapriotti** [comment]
Ah, sorry, I didn't realise there was a type mismatch there. You can fix it by adding the set to list conversion functions in the right place. Here's an excerpt from the schema for `NewConv`:
```haskell
      <*> (Set.toList . newConvAccess)
        .= ( field "access" (Set.fromList <$> array schema)
               <|> pure mempty
           )
```

**Pull request merged**